### PR TITLE
add observability for tsp by prometheus collector

### PR DESCRIPTION
**Pull request by kitianFresh**

### CODE REVIEW

File: pkg/controller/tsp/metric.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
```
Comment (qmhu) [inline pkg/controller/tsp/metric.go:12]: Please combine this line with above import

File: pkg/controller/tsp/metric.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+// Why Implement prometheus collector ？
+// Because the time series prediction timestamp is future timestamp, this way can push timestamp to prometheus
+// if use prometheus metric instrument by default, prometheus scrape will use its own scrape timestamp, so that the prediction time series maybe has wrong timestamps in prom.
+func (c *Controller) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.metric
+}
+
+func (c *Controller) Collect(ch chan<- prometheus.Metric) {
+	// your logic should be placed here
```
Comment (qmhu) [inline pkg/controller/tsp/metric.go:30]: Please remove this comment

File: pkg/controller/tsp/tsp_controller.go [hunk 3]
```diff
@@ -7,6 +7,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (qmhu) [inline pkg/controller/tsp/tsp_controller.go:10]: Also this one

File: pkg/controller/tsp/metric.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocrane/api/prediction/v1alpha1"
+)
+
+func (c *Controller) RegisterMetric() {
```
Comment (mfanjie) [inline pkg/controller/tsp/metric.go:18]: should we have tsp controller implement the metric collector interface, I think we should have a metrics struct instead.

File: pkg/controller/tsp/tsp_controller.go [hunk 5]
```diff
@@ -45,6 +49,12 @@ func NewController(
 		Recorder:     recorder,
 		UpdatePeriod: updatePeriod,
 		predictors:   predictors,
+		metric: prometheus.NewDesc(
```
Comment (mfanjie) [inline pkg/controller/tsp/tsp_controller.go:52]: metrics definition and controller structure should be decouple per my understanding.

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -24,6 +25,16 @@ var (
 		},
 		[]string{"identity", "strategy"},
 	)
+
+	TSPMetric = prometheus.NewGaugeVec(
```
Comment (mfanjie) [inline pkg/metrics/metrics.go:29]: I did not see where is TSPMetric being used

File: pkg/controller/tsp/metric.go (hunk 4 above)
Comment (kitianFresh) [reply pkg/controller/tsp/metric.go:18]: ok, i will refactor it to a metric collector

File: pkg/controller/tsp/tsp_controller.go (hunk 5 above)
Comment (kitianFresh) [reply pkg/controller/tsp/tsp_controller.go:52]: ok

File: pkg/controller/tsp/metric.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocrane/api/prediction/v1alpha1"
```
Comment (wenlxie) [inline pkg/controller/tsp/metric.go:15]: It is better to have a custom package name for this. 
If you want to upgrade the interface, you  just need to change this import package and no need to change the codes.

File: pkg/metrics/tsp_metric_collector.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
```
Comment (qmhu) [inline pkg/metrics/tsp_metric_collector.go:11]: merge the import

File: pkg/metrics/metrics.go [hunk 9]
```diff
@@ -2,6 +2,7 @@ package metrics
 
 import (
 	"github.com/prometheus/client_golang/prometheus"
+
```
Comment (qmhu) [inline pkg/metrics/metrics.go:5]: merge the import

File: cmd/craned/app/manager.go [hunk 10]
```diff
@@ -115,6 +118,11 @@ func Run(ctx context.Context, opts *options.Options) error {
 	return nil
 }
 
+func initializationMetrics(mgr ctrl.Manager) {
```
Comment (qmhu) [inline cmd/craned/app/manager.go:121]: s/initializationMetrics/initializationMetricCollector/

File: pkg/metrics/tsp_metric_collector.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+			string(metricConf.Type),
+			resourceQuery,
+			metricQuery,
+			expressionQuery,
+			string(metricConf.Algorithm.AlgorithmType),
+			key,
+		}
+		samples := data.Samples
+		sort.Slice(samples, func(i, j int) bool {
+			if samples[i].Timestamp < samples[j].Timestamp {
```
Comment (qmhu) [inline pkg/metrics/tsp_metric_collector.go:118]: return samples[i].Timestamp < samples[j].Timestamp

File: pkg/metrics/tsp_metric_collector.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+		})
+
+		// just one timestamp point, because prometheus collector will hash the label values, same label values is not valid
+		for _, sample := range samples {
+			if sample.Timestamp >= now {
+				ts := time.Unix(sample.Timestamp, 0)
+				value, err := strconv.ParseFloat(sample.Value, 64)
+				if err != nil {
+					c.Logger.Error(err, "Failed to parse sample value", "value", value)
+					continue
```
Comment (qmhu) [inline pkg/metrics/tsp_metric_collector.go:132]: better to use exactly nearest sample here

**Pull request merged**